presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPullUnaryNodeAboveProjects.java

```diff
@@ -0,0 +1,166 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.expression;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.markDistinct;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.strictProject;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.facebook.presto.sql.planner.iterative.rule.Util.pullUnaryNodeAboveProjects;
+import static com.facebook.presto.sql.planner.iterative.rule.Util.restrictOutputs;
+import static java.util.Objects.requireNonNull;
+
+public class TestPullUnaryNodeAboveProjects
+        extends BaseRuleTest
+{
+    /**
+     * A rule only for testing the utility function.
+     * PlanMatchPattern needs lots of setup, so it's easier to test the util function as a rule.
+     */
+    private class PullAboveProjects<N extends PlanNode>
+            implements Rule
+    {
+        private Class<N> targetClass;
+
+        PullAboveProjects(Class<N> targetClass)
+        {
+            this.targetClass = requireNonNull(targetClass);
+        }
+
+        @Override
+        public Optional<PlanNode> apply(PlanNode node, Context context)
+        {
+            return pullUnaryNodeAboveProjects(context.getLookup(), targetClass, node)
+                    .flatMap(Optional::<PlanNode>of)
+                    .map(newRoot ->
+                            restrictOutputs(context.getIdAllocator(), newRoot, ImmutableSet.copyOf(node.getOutputSymbols()))
+                                    .orElse(newRoot));
+        }
+    }
+
+    @Test
+    public void testNoTarget()
+    {
+        tester().assertThat(new PullAboveProjects<>(MarkDistinctNode.class))
+                .on(p -> p.values())
+                .doesNotFire();
+    }
+
+    @Test
+    public void testNoProjects()
+    {
+        tester().assertThat(new PullAboveProjects<>(MarkDistinctNode.class))
+                .on(p -> p.markDistinct(p.symbol("marker"), ImmutableList.of(), p.values()))
+                .matches(markDistinct("marker", ImmutableList.of(), values()));
+    }
+
+    @Test
+    public void testTargetCreatedSymbolIdentity()
+    {
+        tester().assertThat(new PullAboveProjects<>(MarkDistinctNode.class))
+                .on(p -> {
+                    Symbol keyUsage = p.symbol("keyUsage");
+                    Symbol marker = p.symbol("marker");
+                    Symbol key = p.symbol("key");
+                    Symbol unused = p.symbol("unused");
+                    return p.project(
+                            Assignments.of(
+                                    marker, marker.toSymbolReference(),
+                                    keyUsage, p.expression("key + 1")),
+                            p.project(
```

Comment:
This is what IntelliJ wants.  If I move the first project call `return p.project(` onto its own line, maybe you'll like the formatting better.  Let's try that!